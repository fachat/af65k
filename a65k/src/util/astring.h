
/****************************************************************************

    string util
    Copyright (C) 2015 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#ifndef STRING_H
#define STRING_H

#include <ctype.h>


// simple hash algorithm. As we use string, we only use the lower 7 bits;
// we also use max the first 8 characters
static inline int string_hash(const char *str) {
	int h = 0;

	int p = 0;
	char c = 0;
	while ( (c = str[p]) ) {
		h = 13 * h + (c & 0x7f);
		p++;
	}
	return h;
}

// simple hash algorithm. As we use string, we only use the lower 7 bits;
// we also use max the first 8 characters
// ignore case by converting tolower() 
static inline int string_hash_nocase(const char *str) {
	int h = 0;

	int p = 0;
	char c = 0;
	while ( (c = str[p]) ) {
		h = 13 * h + (tolower(c) & 0x7f);
		p++;
	}
	return h;
}



#endif

