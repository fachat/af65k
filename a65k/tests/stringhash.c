

#include <stdio.h>

#include "astring.h"


static void test(const char *c) {

	printf("hash of '%s' is %d\n",c, string_hash(c));
}

int main(int argc, char *argv[]) {

	printf("string_hash:\n");

	test("lda");
	test("inx");
	test("nmos_illegal_nobcd");
	test("65802");
	test("65k_w");
	test("abcdefghijklmnopqrstuvwxyz0123456789");
	test("@€_äÖas#*?^↓µ");

	return 0;
}


