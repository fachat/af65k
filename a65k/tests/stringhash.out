string_hash:
hash of 'lda' is 19649
hash of 'inx' is 19295
hash of 'nmos_illegal_nobcd' is -1638109073
hash of '65802' is 1668873
hash of '65k_w' is 1678172
hash of 'abcdefghijklmnopqrstuvwxyz0123456789' is 1200935488
hash of '@€_äÖas#*?^↓µ' is 1226992911
